### PR TITLE
[Book_Store] added shopping cart

### PR DESCRIPTION
**Pull request by fedorovychh**
- created new models
- created their dtos
- created mappers
- created services
- created repositories
- implemented feature than creates shopping cart for user during registration
- created new changesets to liquibase
- created controller for shopping cart with it's endpoints

### CODE REVIEW

File: src/main/java/com/app/bookstore/controller/ShoppingCartController.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    @DeleteMapping("/cart-items/{cartItemId}")
+    @Operation(
+            summary = "Delete book from shopping cart",
+            description = "Delete book from shopping cart by id"
+    )
+    public void deleteItemFromShoppingCart(@PathVariable Long cartItemId) {
+        cartItemService.deleteCartItem(cartItemId);
+    }
+
+    private Long getUserId(Authentication authentication) {
```
Comment (bhdnchui) [inline src/main/java/com/app/bookstore/controller/ShoppingCartController.java]: It's unnecessary to make a separate method if you use it one time.

Comment (bhdnchui) [inline src/main/java/com/app/bookstore/controller/ShoppingCartController.java]: It's unnecessary to make a separate method if you use it one time.

File: src/main/java/com/app/bookstore/dto/cart/item/AddCartItemRequestDto.java [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package com.app.bookstore.dto.cart.item;
+
+import lombok.Data;
+
+@Data
+public class AddCartItemRequestDto {
+    private Long bookId;
+    private Integer quantity;
```
Comment (bhdnchui) [inline src/main/java/com/app/bookstore/dto/cart/item/AddCartItemRequestDto.java]: Add constraints here.

File: src/main/java/com/app/bookstore/repository/cart/item/CartItemRepository.java [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+import java.util.List;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface CartItemRepository extends JpaRepository<CartItem, Long> {
+    @Query("SELECT b FROM CartItem b JOIN FETCH b.shoppingCart")
```
Comment (bhdnchui) [inline src/main/java/com/app/bookstore/repository/cart/item/CartItemRepository.java]: Why do you use `b` here?

File: src/main/java/com/app/bookstore/repository/cart/item/CartItemRepository.java [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+    Page<CartItem> findAll(Pageable pageable);
+
+    @Query("SELECT b FROM CartItem b JOIN fetch b.shoppingCart c WHERE c.id = :itemId")
```
Comment (bhdnchui) [inline src/main/java/com/app/bookstore/repository/cart/item/CartItemRepository.java]: One random letter naming

File: src/main/java/com/app/bookstore/service/cart/item/CartItemServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        CartItem cartItem = cartItemMapper.toCartItem(requestDto);
+        cartItem.setShoppingCart(shoppingCart);
+        Long bookId = requestDto.getBookId();
+        cartItem.setBook(bookRepository.findById(bookId).orElseThrow(
+                () -> new EntityNotFoundException("Can't find book by id: " + bookId)
+        ));
+        return cartItemRepository.save(cartItem);
+    }
+
+    private CartItem parseCartItem(Long cartItemId) {
```
Comment (bhdnchui) [inline src/main/java/com/app/bookstore/service/cart/item/CartItemServiceImpl.java]: Rename. You did not parse anything here.

File: src/main/java/com/app/bookstore/service/cart/item/CartItemServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+@Service
+@RequiredArgsConstructor
+public class CartItemServiceImpl implements CartItemService {
+    private final CartItemRepository cartItemRepository;
+    private final CartItemMapper cartItemMapper;
+    private final BookRepository bookRepository;
+
+    @Override
+    public CartItemResponseDto updateCartItemById(Long cartItemId, int quantity) {
```
Comment (bhdnchui) [inline src/main/java/com/app/bookstore/service/cart/item/CartItemServiceImpl.java]: If it is CartItemService, than we can avoid using CardItem in methods.
```suggestion
    public CartItemResponseDto updateById(Long id, int quantity) {

```

Comment (bhdnchui) [reply src/main/java/com/app/bookstore/service/cart/item/CartItemServiceImpl.java]: Make sure to simplify it everywhere.

File: src/main/java/com/app/bookstore/service/shopping/cart/ShoppingCartService.java [hunk 7]
```diff
@@ -0,0 +1,22 @@
+package com.app.bookstore.service.shopping.cart;
+
+import com.app.bookstore.dto.cart.item.PutCartItemRequestDto;
+import com.app.bookstore.dto.shopping.cart.ShoppingCartRequestDto;
+import com.app.bookstore.dto.shopping.cart.ShoppingCartResponseDto;
+import com.app.bookstore.model.User;
+import org.springframework.security.core.Authentication;
+
+public interface ShoppingCartService {
+    ShoppingCartResponseDto findShoppingCartByUserId(Long userId);
```
Comment (bhdnchui) [inline src/main/java/com/app/bookstore/service/shopping/cart/ShoppingCartService.java]: Same as for CardItemService.
```suggestion
    ShoppingCartResponseDto findByUserId(Long userId);
```

Comment (bhdnchui) [reply src/main/java/com/app/bookstore/service/shopping/cart/ShoppingCartService.java]: Make sure to simplify it everywhere.

File: src/main/java/com/app/bookstore/repository/shopping/cart/ShoppingCartRepository.java [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+
+import com.app.bookstore.model.ShoppingCart;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface ShoppingCartRepository extends JpaRepository<ShoppingCart, Long> {
+    @Query("SELECT sc FROM ShoppingCart sc JOIN FETCH sc.cartItems ci WHERE sc.user.id = :userId")
+    ShoppingCart findShoppingCartByUserId(Long userId);
```
Comment (bhdnchui) [inline src/main/java/com/app/bookstore/repository/shopping/cart/ShoppingCartRepository.java]: Return `Optional` here

**Review by bhdnchui — CHANGES_REQUESTED** [review]
A few points to consider.

**Review by vovasalyha — APPROVED** [review]

**Pull request merged**